File: packages/sdk-python/src/pr_sdk/http.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional

import httpx

from .errors import (
    PrNetworkError,
    PrServerError,
    PrTimeoutError,
    create_error_for_status,
    parse_error_body,
)

DEFAULT_BASE_URL = "http://localhost:3001"
DEFAULT_TIMEOUT_SECONDS = 10.0


@dataclass
class RequestContext:
    base_url: str
    api_key: str
    timeout_seconds: float


def build_request_url(context: RequestContext, path: str) -> str:
    return context.base_url + path


def _parse_retry_after_ms(response: httpx.Response) -> Optional[int]:
    header = response.headers.get("retry-after")

    if header is None:
        return None

    try:
        return max(int(float(header) * 1000), 0)
    except ValueError:
        return None


def _handle_response(response: httpx.Response) -> Any:
    if 200 <= response.status_code < 300:
        if not response.content:
            return None

        return response.json()

    parsed = parse_error_body(response.text)
    retry_after_ms = _parse_retry_after_ms(response) if response.status_code == 429 else None

    raise create_error_for_status(
        response.status_code,
        parsed.message,
        code=parsed.code,
        issues=parsed.issues,
        retry_after_ms=retry_after_ms,
    )


def request_sync(
    context: RequestContext,
    client: httpx.Client,
    path: str,
    *,
    method: str = "GET",
    query: Optional[Dict[str, str]] = None,
    json_body: Optional[Dict[str, Any]] = None,
) -> Any:
    try:
        response = client.request(
            method,
            build_request_url(context, path),
            params={k: v for k, v in query.items() if v is not None} if query else None,
            json=json_body if json_body is not None else None,
            headers={"Authorization": f"Bearer {context.api_key}"},
        )
    except httpx.TimeoutException as error:
        raise PrTimeoutError(f"request timed out after {context.timeout_seconds}s") from error
    except httpx.HTTPError as error:
        raise PrNetworkError("network request failed") from error

    return _handle_response(response)


async def request_async(
    context: RequestContext,
    client: httpx.AsyncClient,
    path: str,
    *,
    method: str = "GET",
    query: Optional[Dict[str, str]] = None,
    json_body: Optional[Dict[str, Any]] = None,
) -> Any:
    import asyncio

    try:
        response = await client.request(
            method,
            build_request_url(context, path),
            params={k: v for k, v in query.items() if v is not None} if query else None,
            json=json_body if json_body is not None else None,
            headers={"Authorization": f"Bearer {context.api_key}"},
        )
    except asyncio.CancelledError:
        raise
    except httpx.TimeoutException as error:
        raise PrTimeoutError(f"request timed out after {context.timeout_seconds}s") from error
    except httpx.HTTPError as error:
        raise PrNetworkError("network request failed") from error

    return _handle_response(response)


__all__ = [
    "DEFAULT_BASE_URL",
    "DEFAULT_TIMEOUT_SECONDS",
    "RequestContext",
    "request_sync",
    "request_async",
    "PrServerError",
]

File: packages/sdk-typescript/src/http.ts
import {
  PrNetworkError,
  PrTimeoutError,
  createPrErrorForStatus,
  parseErrorBody,
} from "./errors.js";

export type HttpMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE";

export type RequestOptions = {
  method?: HttpMethod;
  query?: Record<string, string | undefined>;
  body?: unknown;
  signal?: AbortSignal;
};

export type RequestContext = {
  baseUrl: string;
  apiKey: string;
  timeoutMs: number;
  fetchImpl: typeof fetch;
};

function buildUrl(baseUrl: string, path: string, query: Record<string, string | undefined> | undefined): string {
  const url = new URL(path, baseUrl);

  if (query) {
    for (const [key, value] of Object.entries(query)) {
      if (value !== undefined) {
        url.searchParams.set(key, value);
      }
    }
  }

  return url.toString();
}

function createAbortError(externalSignal: AbortSignal | undefined): unknown {
  const reason = externalSignal?.reason;

  if (reason !== undefined) {
    return reason;
  }

  return new DOMException("This operation was aborted", "AbortError");
}

function bindExternalSignal(
  externalSignal: AbortSignal | undefined,
  controller: AbortController,
): () => void {
  if (!externalSignal) {
    return () => {};
  }

  const listener = () => controller.abort();
  externalSignal.addEventListener("abort", listener, { once: true });

  return () => externalSignal.removeEventListener("abort", listener);
}

async function readResponseBody(response: Response): Promise<unknown> {
  const text = await response.text();

  if (text === "") {
    return null;
  }

  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

export async function request<T>(
  context: RequestContext,
  path: string,
  options: RequestOptions = {},
): Promise<T> {
  const { baseUrl, apiKey, timeoutMs, fetchImpl } = context;
  const method = options.method ?? "GET";
  const url = buildUrl(baseUrl, path, options.query);

  if (options.signal?.aborted) {
    throw createAbortError(options.signal);
  }

  // Combine the caller's signal with our own timeout without relying solely on
  // AbortSignal.timeout(), so caller cancellation still works.
  const controller = new AbortController();
  let timedOut = false;

  const timeoutId = setTimeout(() => {
    timedOut = true;
    controller.abort();
  }, timeoutMs);

  const unbindExternalSignal = bindExternalSignal(options.signal, controller);

  try {
    let response: Response;
    let responseBody: unknown;

    try {
      response = await fetchImpl(url, {
        method,
        headers: {
          Authorization: `Bearer ${apiKey}`,
          ...(options.body !== undefined ? { "Content-Type": "application/json" } : {}),
        },
        body: options.body !== undefined ? JSON.stringify(options.body) : undefined,
        signal: controller.signal,
      });
      responseBody = await readResponseBody(response);
    } catch (error) {
      if (timedOut) {
        throw new PrTimeoutError(`request timed out after ${timeoutMs}ms`);
      }

      if (options.signal?.aborted) {
        throw createAbortError(options.signal);
      }

      throw new PrNetworkError("network request failed", { cause: error });
    }

    if (!response.ok) {
      const parsed = parseErrorBody(responseBody);

      throw createPrErrorForStatus(response.status, parsed.message, {
        code: parsed.code,
        issues: parsed.issues,
        ...(response.status === 429 ? { retryAfterMs: parseRetryAfterMs(response.headers.get("retry-after")) } : {}),
      });
    }

    return responseBody as T;
  } finally {
    clearTimeout(timeoutId);
    unbindExternalSignal();
  }
}

function parseRetryAfterMs(headerValue: string | null): number | undefined {
  if (headerValue === null) {
    return undefined;
  }

  const seconds = Number(headerValue);

  if (!Number.isFinite(seconds)) {
    return undefined;
  }

  return Math.max(seconds * 1000, 0);
}
